platform/android/export/export.cpp

```diff
@@ -2113,43 +2162,37 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 			//write
 
 			if (file == "AndroidManifest.xml") {
-
 				_fix_manifest(p_preset, data, p_flags & (DEBUG_FLAG_DUMB_CLIENT | DEBUG_FLAG_REMOTE_DEBUG));
 			}
 
 			if (file == "resources.arsc") {
-
 				_fix_resources(p_preset, data);
 			}
 
-			if (file == "res/drawable-nodpi-v4/icon.png") {
-				bool found = false;
-				for (uint64_t i = 0; i < sizeof(launcher_icons) / sizeof(launcher_icons[0]); ++i) {
-					String icon_path = String(p_preset->get(launcher_icons[i].option_id)).strip_edges();
-					if (icon_path != "" && icon_path.ends_with(".png")) {
-						FileAccess *f = FileAccess::open(icon_path, FileAccess::READ);
-						if (f) {
-							data.resize(f->get_len());
-							f->get_buffer(data.ptrw(), data.size());
-							memdelete(f);
-							found = true;
-							break;
-						}
-					}
-				}
-				if (!found) {
-
-					String appicon = ProjectSettings::get_singleton()->get("application/config/icon");
-					if (appicon != "" && appicon.ends_with(".png")) {
-						FileAccess *f = FileAccess::open(appicon, FileAccess::READ);
-						if (f) {
-							data.resize(f->get_len());
-							f->get_buffer(data.ptrw(), data.size());
-							memdelete(f);
-						}
-					}
-				}
+			String project_icon_path = ProjectSettings::get_singleton()->get("application/config/icon");
+
+			Vector<String> fallback_icons;
+			fallback_icons.push_back(project_icon_path);
+			for (uint64_t i = 0; i < sizeof(launcher_icons) / sizeof(launcher_icons[0]); ++i) {
+				_process_launcher_icons(p_preset, file, launcher_icons[i], data, fallback_icons);
+			}
+			//Replace the non-specific density icon for completeness, it should almost never be used by the system.
+			_process_launcher_icons(p_preset, file, { "launcher_icons/xxxhdpi_192x192", "res/mipmap/icon.png" }, data, fallback_icons);
+
+			fallback_icons.clear();
+			fallback_icons.push_back("");
+			fallback_icons.push_back(project_icon_path);
+			for (uint64_t i = 0; i < sizeof(launcher_adaptive_icons_foreground) / sizeof(launcher_adaptive_icons_foreground[0]); ++i) {
+				fallback_icons.set(0, String(p_preset->get(launcher_icons[i].option_id)).strip_edges()); //Regular icon as fallback to adaptive foreground
+				_process_launcher_icons(p_preset, file, launcher_adaptive_icons_foreground[i], data, fallback_icons);
```

Comment:
The logic here and below seems to assume that the user will fill all of the icon density entries; is that guaranteed?
This is a regression for the current logic which just checks if at least one icon is specified, and use that.

With the new logic, if the user only specifies one icon density, the fallbacks will be set for the other icon densities leading to a mismatch of icons on devices of different density.

Reply:
Yes, the mismatch between densities can happen. Requesting a single image and resizing it should solve this problem.